pkgs/development/tools/build-managers/bloop/default.nix

```
@@ -1,78 +1,92 @@
+    sha256 = "1ldxlqv353gvhdn4yq7z506ywvnjv6fjsi8wigwhzg89876pwsys";
   };
 
-  server = stdenv.mkDerivation {
-    name = "${baseName}-server-${version}";
-    buildCommand = ''
-      mkdir -p $out/bin
+  bloop-fish = fetchurl {
+    url = "https://github.com/scalacenter/bloop/releases/download/v${version}/fish-completions";
+    sha256 = "1pa8h81l2498q8dbd83fzipr99myjwxpy8xdgzhvqzdmfv6aa4m0";
+  };
 
-      export COURSIER_CACHE=$(pwd)
-      ${coursier}/bin/coursier bootstrap ch.epfl.scala:bloop-frontend_2.12:${version} \
-        -r "bintray:scalameta/maven" \
-        -r "bintray:scalacenter/releases" \
-        -r "https://oss.sonatype.org/content/repositories/staging" \
-        --deterministic \
-        -f --main bloop.Server -o $out/bin/blp-server
-    '';
-    outputHashMode = "recursive";
-    outputHashAlgo = "sha256";
-    outputHash     = "1z33ip6hgfwiixm2gimz819p5cnxn1fmxb3ryyf77jzwsx7py718";
+  bloop-zsh = fetchurl {
+    url = "https://github.com/scalacenter/bloop/releases/download/v${version}/zsh-completions";
+    sha256 = "1xzg0qfkjdmzm3mvg82mc4iia8cl7b6vbl8ng4ir2xsz00zjrlsq";
   };
 
-  zsh = stdenv.mkDerivation {
-    name = "${baseName}-zshcompletion-${version}";
+  bloop-coursier = stdenv.mkDerivation rec {
+    name = "${pname}-coursier-${version}";
 
-    src = fetchurl {
-      url = "https://raw.githubusercontent.com/scalacenter/bloop/v${version}/etc/zsh/_bloop";
-      sha256 = "09qq5888vaqlqan2jbs2qajz2c3ff13zj8r0x2pcxsqmvlqr02hp";
-    };
+    platform = if stdenv.isLinux && stdenv.isx86_64 then "x86_64-pc-linux"
+               else if stdenv.isDarwin && stdenv.isx86_64 then "x86_64-apple-darwin"
+               else throw "unsupported platform";
 
     phases = [ "installPhase" ];
+    installPhase = ''
+      export COURSIER_CACHE=$(pwd)
+      export COURSIER_JVM_CACHE=$(pwd)
 
-    installPhase = ''cp $src $out'';
-  };
-in
-stdenv.mkDerivation {
-  name = "${baseName}-${version}";
+      mkdir channel
+      ln -s ${bloop-coursier-channel} channel/bloop.json
+      ${coursier}/bin/coursier install --install-dir $out --install-platform ${platform} --default-channels=false --channel channel --only-prebuilt=true bloop
 
-  buildInputs = [ makeWrapper ];
+      # Remove binary part of the coursier launcher script to make derivation output hash stable
+      sed -i '5,$ d' $out/bloop
+   '';
 
-  phases = [ "installPhase" ];
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+    outputHash     = if stdenv.isLinux && stdenv.isx86_64 then "1ncl34f39mvk0zb5jl1l77cwjdg3xfnhjxbzz11pdfqw0d7wqywj"
+                     else if stdenv.isDarwin && stdenv.isx86_64 then "06c885w088yvh8l1r1jbrz0549gx2xvc8xr6rlxy6y27jk5655p2"
+                     else throw "unsupported platform";
+  };
+
+  nativeBuildInputs = [ autoPatchelfHook ] ;
+  phases = [ "installPhase" "fixupPhase" ];
+  buildInputs = [ stdenv.cc.cc.lib zlib ] ;
```

fix java missing issue
```suggestion
  buildInputs = [ stdenv.cc.cc.lib zlib ] ;
  propagatedBuildInputs = [ openjdk.jre ];
```

---

this does increase the closure size quite a bit:
```
[nix-shell:/home/jon/.cache/nixpkgs-review/pr-91758-4]$ nix path-info -Sh ./results/bloop
/nix/store/fp1kc7lxw0rhhs6my624wm154hn8494w-bloop-1.4.3	  77.2M

[nix-shell:/home/jon/.cache/nixpkgs-review/pr-91758-4]$ nix path-info -Sh ./result
/nix/store/c0ls7dh4vnjd4wyqib3mz6yyfz2n3g49-bloop-1.4.3	 434.8M
```
However, it's more representative of it's actual runtime closure:
```
[16:52:28] jon@jon-desktop /home/jon/projects/nixpkgs (buildDotnetCorePackage)
$ nix path-info -Sh ./result-jre
/nix/store/7qknskbanpwq7a71s2n6dv3l5b3frj7d-openjdk-8u242-b08-jre	 396.5M
```

---

This does still mean that we can change to any java I want when using bloop, right? :)

In fact, why are we adding java 8? Does the build not work without it or usage?

---

Looks like a similar thing is done for sbt already - https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/sbt/default.nix, and we can override jvm for bloop processes, so yeah (answering my own question)  :)

---

Would be good if there was a way to configure preferred java provider in one place though, but thats much more generic question than this PR

---

as with most things with nix, there's a way to do anything, it's just how much pain is it to introduce flexibility. I guess you could pass `jre` to this derivation, and then just say:
```
  bloop = callPackage ../.. {
    jre = openjdkXX.jre;
  };
```
and then if you want to change it, then you would do something like:
```
  bloop.override { jre = blah.jre; };
```

---

It does not seem necessary to explicitly pass the jre in the callPackage call, several other derivations which use jre do not do it (coursier, scalafmt, metals, ...).

However by passing the jre inside the derivation, it seems to respect the programs.java.package property or can be overriden with derivation.override.